iree/compiler/Dialect/HAL/IR/test/tensor_op_folding.mlir

```diff
@@ -0,0 +1,33 @@
+// RUN: iree-opt -split-input-file -canonicalize -cse %s | iree-opt -split-input-file | IreeFileCheck %s
+
+// CHECK-LABEL: @tensorCastMatchingTypeFolds
+func @tensorCastMatchingTypeFolds(%arg0: !hal.buffer_view) -> !hal.buffer_view {
+  // CHECK-NOT: hal.tensor.cast
+  // CHECK: return %arg0 : !hal.buffer_view
+  %0 = hal.tensor.cast %arg0 : !hal.buffer_view -> !hal.buffer_view
+  return %0 : !hal.buffer_view
+}
+
+// -----
+
+// CHECK-LABEL: @tensorCastPassthroughFolds
+func @tensorCastPassthroughFolds(%arg0: !hal.buffer_view) -> !hal.buffer_view {
+  // CHECK-NOT: hal.tensor.cast
+  // CHECK: return %arg0 : !hal.buffer_view
+  %0 = hal.tensor.cast %arg0 : !hal.buffer_view -> tensor<5xi32>
+  %1 = hal.tensor.cast %0 : tensor<5xi32> -> !hal.buffer_view
+  return %1 : !hal.buffer_view
+}
+
+// -----
+
+// CHECK-LABEL: @tensorCastDifferentTypesDoNotFold
+func @tensorCastDifferentTypesDoNotFold(%arg0: !hal.buffer_view) -> !hal.buffer {
+  // This could be canonicalized if there is only one use.
+  // CHECK: %0 = hal.tensor.cast %arg0 : !hal.buffer_view -> tensor<5xi32>
+  // CHECK-NEXT: %1 = hal.tensor.cast %0 : tensor<5xi32> -> !hal.buffer
+  // CHECK-NEXT: return %1 : !hal.buffer
+  %0 = hal.tensor.cast %arg0 : !hal.buffer_view -> tensor<5xi32>
+  %1 = hal.tensor.cast %0 : tensor<5xi32> -> !hal.buffer
```

Comment:
Could implement this as part of this PR too?

Reply:
That'd be nice! These now show up in all our paths for pass-through things, and canonicalizing them early should avoid ever creating the stream.tensor.import/export ops in the first place.

Reply:
Yeah, the `hal.tensor.cast` folder works for the passthrough ops I've seen:
```mlir
  %0 = hal.tensor.cast %arg0 : !hal.buffer_view -> tensor<5xi32>
  %1 = hal.tensor.cast %0 : tensor<5xi32> -> !hal.buffer_view
```
(A -> B, B -> A). That does avoid creating the stream ops.

However, it doesn't apply to this, which I haven't seen (yet?):
```mlir
  %0 = hal.tensor.cast %arg0 : !hal.buffer_view -> tensor<5xi32>
  %1 = hal.tensor.cast %0 : tensor<5xi32> -> !hal.buffer
```
(A -> B, B -> C)

Reply:
Implemented within the folder 👍